### PR TITLE
NCN Automation Business Logic

### PR DESCRIPTION
**Pull request by SeanWallace**
#### Summary and Scope

This mod contains all of the changes to CSI necessary to support all the business logic associated with NCN lifecycle events. And it's freaking huge...I'm sorry for that. 95% of the changes is vendor code so obviously that can be ignored. 

The rest is almost purely additive, as in it shouldn't be possible for it to break anything all by itself. For the files that I did touch that already existed that was almost exclusively for reorganization purposes to break out shared functions into common places.

- Relates to # CASMINST-3029

##### Issue Type

- RFE Pull Request

#### Prerequisites

- [x] I have included documentation in my PR (or it is not required)
- [x] I tested this on internal system (x) (if yes, please include results or a description of the test)
 
#### Idempotency
 

All of this code is inherently idempotent in that it all seeks to get the selected resource to a desired state.
 
#### Risks and Mitigations
 

As mentioned above this should be pretty low risk, majority is new.

### CODE REVIEW

File: cmd/handoff-bss-metadata.go [hunk 1]
```diff
@@ -51,9 +51,9 @@ var (
 	dataFile      string
 	cloudInitData map[string]bssTypes.CloudInit
 	sshConfig     *ssh.ClientConfig
-	token         string
 
-	vlansToGather = []string{"bond0.nmn0"}
+	// TODO: SET THIS BACK
+	vlansToGather = []string{"vlan002"}
```
Comment (jacobsalmela) [inline cmd/handoff-bss-metadata.go:56]: Does his needs to be changed back now to work with 1.2?

**Review by jacobsalmela — COMMENTED** [review]
Looks mergeable.  Just that one question I had in the comment.

**Comment by rustydb** [comment]
Sean, I'm still reviewing this. I just haven't had time yet, I should be done tonight or in the morning.
I am just mainly keen on looking at the new build stuff and go1.16 before this merges.

**Comment by SeanWallace** [comment]
> Sean, I'm still reviewing this. I just haven't had time yet, I should be done tonight or in the morning.
> I am just mainly keen on looking at the new build stuff and go1.16 before this merges.

You're good, I'm so far behind on other things I haven't had time to even loop back on all the rest of the PRs I've opened yet. Appreciate the review. :)

**Comment by rustydb** [comment]
> > Sean, I'm still reviewing this. I just haven't had time yet, I should be done tonight or in the morning.
> > I am just mainly keen on looking at the new build stuff and go1.16 before this merges.
> 
> You're good, I'm so far behind on other things I haven't had time to even loop back on all the rest of the PRs I've opened yet. Appreciate the review. :)
> 
> Oh, and I need to still fix the way this Docker build works, this build methodology causes every file to be owned by root in the build path and that makes the pipeline barf.

I was wondering about the Docker build, do we need to build an image _in_ this repo? The `Dockerfile` seems to be very not-CSI specific. I think we should move it to csm-build-environments, and then just pull it via the `Jenkinsfile`:
```groovy
  agent {
    docker {
        label "metal-gcp-builder"
        image "https://artifactory.algol60.net/artifactory/csm-docker/stable/sle15sp3_golang_build_environment
:latest"
        args "-u root -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker --group-add 999"
    }
  }
```

**Comment by rustydb** [comment]
> > > Sean, I'm still reviewing this. I just haven't had time yet, I should be done tonight or in the morning.
> > > I am just mainly keen on looking at the new build stuff and go1.16 before this merges.
> > 
> > 
> > You're good, I'm so far behind on other things I haven't had time to even loop back on all the rest of the PRs I've opened yet. Appreciate the review. :)
> > Oh, and I need to still fix the way this Docker build works, this build methodology causes every file to be owned by root in the build path and that makes the pipeline barf.
> 
> I was wondering about the Docker build, do we need to build an image _in_ this repo? The `Dockerfile` seems to be very not-CSI specific. I think we should move it to csm-build-environments, and then just pull it via the `Jenkinsfile`: (imagining if you called this image "sle15sp3_golang_build_environment" this is what the Groovy code would look like)
> 
> ```groovy
>   agent {
>     docker {
>         label "metal-gcp-builder"
>         image "https://artifactory.algol60.net/artifactory/csm-docker/stable/sle15sp3_golang_build_environment
> :latest"
>         args "-u root -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker --group-add 999"
>     }
>   }
> ```
> 
> Since it seems that the Docker image is portable and could be used elsewhere, it doesn't make sense to live in this repo. I understand using it here for development/testing, but definitely moving it would make this build even easier.. since then you'd just pull and build.

For example, you could make a `sles15sp3_go` directory in https://github.com/Cray-HPE/csm-build-environments/ with your `Dockerfile`. It'll push to algol60 and then it can be available for anyone else (without depending on the CSI repo).

**Comment by SeanWallace** [comment]
> > > > Sean, I'm still reviewing this. I just haven't had time yet, I should be done tonight or in the morning.
> > > > I am just mainly keen on looking at the new build stuff and go1.16 before this merges.
> > > 
> > > 
> > > You're good, I'm so far behind on other things I haven't had time to even loop back on all the rest of the PRs I've opened yet. Appreciate the review. :)
> > > Oh, and I need to still fix the way this Docker build works, this build methodology causes every file to be owned by root in the build path and that makes the pipeline barf.
> > 
> > 
> > I was wondering about the Docker build, do we need to build an image _in_ this repo? The `Dockerfile` seems to be very not-CSI specific. I think we should move it to csm-build-environments, and then just pull it via the `Jenkinsfile`: (imagining if you called this image "sle15sp3_golang_build_environment" this is what the Groovy code would look like)
> > ```groovy
> >   agent {
> >     docker {
> >         label "metal-gcp-builder"
> >         image "https://artifactory.algol60.net/artifactory/csm-docker/stable/sle15sp3_golang_build_environment
> > :latest"
> >         args "-u root -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker --group-add 999"
> >     }
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since it seems that the Docker image is portable and could be used elsewhere, it doesn't make sense to live in this repo. I understand using it here for development/testing, but definitely moving it would make this build even easier.. since then you'd just pull and build.
> 
> For example, you could make a `sles15sp3_go` directory in https://github.com/Cray-HPE/csm-build-environments/ with your `Dockerfile`. It'll push to algol60 and then it can be available for anyone else (without depending on the CSI repo).

I can certainly move this over there the only thing I ever worry about doing something like that is there is nothing linking the two of these things together then. What I mean by that is somebody might make a very incompatible change over in that other repo that effectively breaks this image and nobody maintaining this repo would be any the wiser until the builds started failing. 

The other thing is to me it doesn't look as simple as just defining a new directory and moving this Dockerfile over there. Looking at the current [Jenkinsfile](https://github.com/Cray-HPE/csm-build-environments/blob/main/Jenkinsfile) it looks super specific to the sp3 build environment. I'm sure it's possible to make a new build job and and overcome this issue.

I guess my point is, sure, there's nothing specific about this Dockerfile to CSI at least for right now. As a result is it nice to imagine it lives in a more universal place? Yeah, totally. But if it ever did need to become specific to CSI for any reason either it would need to move back here or the base image would need to be inherited from and modified either of which requiring a Dockerfile. 

I'm not against it by any means I'm also just not convinced it makes anything easier or more maintainable. If you feel that strongly about it I'll break it out and work with Michael to get it built and tagged.

File: cmd/handoff-bss-metadata.go (hunk 1 above)
Comment (SeanWallace) [reply cmd/handoff-bss-metadata.go:56]: GREAT catch thank you so much for finding that!

**Review by heemstra — APPROVED** [review]

**Review by heemstra — APPROVED** [review]

**Comment by rustydb** [comment]
> > > > > Sean, I'm still reviewing this. I just haven't had time yet, I should be done tonight or in the morning.
> > > > > I am just mainly keen on looking at the new build stuff and go1.16 before this merges.
> > > > 
> > > > 
> > > > You're good, I'm so far behind on other things I haven't had time to even loop back on all the rest of the PRs I've opened yet. Appreciate the review. :)
> > > > Oh, and I need to still fix the way this Docker build works, this build methodology causes every file to be owned by root in the build path and that makes the pipeline barf.
> > > 
> > > 
> > > I was wondering about the Docker build, do we need to build an image _in_ this repo? The `Dockerfile` seems to be very not-CSI specific. I think we should move it to csm-build-environments, and then just pull it via the `Jenkinsfile`: (imagining if you called this image "sle15sp3_golang_build_environment" this is what the Groovy code would look like)
> > > ```groovy
> > >   agent {
> > >     docker {
> > >         label "metal-gcp-builder"
> > >         image "https://artifactory.algol60.net/artifactory/csm-docker/stable/sle15sp3_golang_build_environment
> > > :latest"
> > >         args "-u root -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker --group-add 999"
> > >     }
> > >   }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Since it seems that the Docker image is portable and could be used elsewhere, it doesn't make sense to live in this repo. I understand using it here for development/testing, but definitely moving it would make this build even easier.. since then you'd just pull and build.
> > 
> > 
> > For example, you could make a `sles15sp3_go` directory in https://github.com/Cray-HPE/csm-build-environments/ with your `Dockerfile`. It'll push to algol60 and then it can be available for anyone else (without depending on the CSI repo).
> 
> I can certainly move this over there the only thing I ever worry about doing something like that is there is nothing linking the two of these things together then. What I mean by that is somebody might make a very incompatible change over in that other repo that effectively breaks this image and nobody maintaining this repo would be any the wiser until the builds started failing.
> 
> The other thing is to me it doesn't look as simple as just defining a new directory and moving this Dockerfile over there. Looking at the current [Jenkinsfile](https://github.com/Cray-HPE/csm-build-environments/blob/main/Jenkinsfile) it looks super specific to the sp3 build environment. I'm sure it's possible to make a new build job and and overcome this issue.
> 
> I guess my point is, sure, there's nothing specific about this Dockerfile to CSI at least for right now. As a result is it nice to imagine it lives in a more universal place? Yeah, totally. But if it ever did need to become specific to CSI for any reason either it would need to move back here or the base image would need to be inherited from and modified either of which requiring a Dockerfile.
> 
> I'm not against it by any means I'm also just not convinced it makes anything easier or more maintainable. If you feel that strongly about it I'll break it out and work with Michael to get it built and tagged.

Maybe I don't follow, it looked like you made a generic GO container.

**Comment by rustydb** [comment]
I'll make a JIRA for moving the new build step and new added bulk out of the repo, since I wanted to use the new generic go container for basecamp.

File: Makefile [hunk 2]
```diff
@@ -164,3 +171,11 @@ version:
 update-version: build
 	@echo 'Version = ${CHANGELOG_VERSION}'
 	echo ${CHANGELOG_VERSION} > .version
+
+docker-build-image:
+	docker build --build-arg GO_VERSION=$(GO_VERSION) \
+	-t cray/csi-rpm-builder .
+
+docker-rpm: docker-build-image
+	docker run --user $(CURRENT_UID):$(CURRENT_GID) --env GIT_REPO_NAME=cray-site-init --rm -v $(CURDIR):/build \
```
Comment (rustydb) [inline Makefile:180]: I'm struggling to recall, does docker implicitly know which image to use?

Comment (SeanWallace) [reply Makefile:180]: It doesn't, the next line says `cray/csi-rpm-builder` which is the image.

**Pull request merged**